nemo/collections/nlp/models/language_modeling/megatron/tokens_encoder_decoder.py

```diff
@@ -0,0 +1,282 @@
+# Copyright (c) 2021, NVIDIA CORPORATION.  All rights reserved.
+#
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+#     http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+
+"""T5 model."""
+
+import torch
+
+from nemo.collections.nlp.modules.common.megatron.language_model import Embedding
+from nemo.collections.nlp.modules.common.megatron.megatron_decoders import get_decoder_model
+from nemo.collections.nlp.modules.common.megatron.megatron_encoder_decoder import (
+    MegatronTransformerEncoderDecoderModule,
+)
+from nemo.collections.nlp.modules.common.megatron.megatron_encoders import get_encoder_model
+from nemo.collections.nlp.modules.common.megatron.module import MegatronModule
+from nemo.collections.nlp.modules.common.megatron.utils import (
+    build_position_ids,
+    init_method_normal,
+    parallel_lm_logits,
+    scaled_init_method_normal,
+)
+
+try:
+    from apex.transformer import tensor_parallel
+    from apex.transformer.enums import AttnMaskType
+
+    HAVE_APEX = True
+except (ImportError, ModuleNotFoundError):
+    HAVE_APEX = False
+
+
+__all__ = ["MegatronTokensHead", "TokensEncoderDecoderModule"]
+
+
+class MegatronTokensHead(MegatronModule):
+    """Masked LM head for token-based encoder-decoder models (e.g., T5)
+
+    Arguments:
+        mpu_vocab_size: model parallel size of vocabulary.
+        parallel_output: wether output logits being distributed or not.
+    """
+
+    def __init__(self, mpu_vocab_size, parallel_output):
+        super(MegatronTokensHead, self).__init__()
+
+        self.bias = torch.nn.Parameter(torch.zeros(mpu_vocab_size))
+        self.bias.model_parallel = True
+        self.bias.partition_dim = 0
+        self.bias.stride = 1
+        self.parallel_output = parallel_output
+
+    def forward(self, hidden_states, word_embeddings_weight):
+        output = parallel_lm_logits(hidden_states, word_embeddings_weight, self.parallel_output, bias=self.bias)
+        return output
+
+
+# TODO: add soft prompts
+
+
+class TokensEncoderDecoderModule(MegatronModule):
```

Comment:
Should this be in the modules folder?

Reply:
Worth discussion. Could either stay here (being specific to tokens, or moved to modules being rather generic.
@MaximumEntropy, @ericharper, @okuchaiev ?

Reply:
Done.